Detectors/ZDC/fastsimulation/include/FastSimulations.hpp

```diff
@@ -0,0 +1,99 @@
+/**
+ * @file FastSimulations.hpp
+ * @author SwirtaB
+ * @brief
+ * @version 0.1.2
+ * @date 2021-12-20
+ *
+ * @copyright Copyright (c) 2021
+ *
+ */
+
+#ifndef ONNX_API_FAST_SIMULATIONS_HPP
+#define ONNX_API_FAST_SIMULATIONS_HPP
+
+#include "Config.hpp"
+#include <onnxruntime/core/session/onnxruntime_cxx_api.h>
+#include <array>
+#include <optional>
+
+namespace fs
+{
+std::array<int, 5> calculate_channels(Ort::Value& value);
+
+class NeuralFastSimulation
+{
+ public:
+  NeuralFastSimulation(const std::string& modelPath,
+                       Ort::SessionOptions sessionOptions,
+                       OrtAllocatorType allocatorType,
+                       OrtMemType memoryType);
+  ~NeuralFastSimulation() = default;
+
+  virtual void run() = 0;
+  virtual std::array<int, 5> get_channels() = 0;
+  // virtual void               reload_scales(const std::string &path) = 0;
+
+ protected:
+  void set_input_output_data();
+
+  /// ONNX specific attributes
+  Ort::Env m_env;
+  Ort::Session m_session;
+  Ort::AllocatorWithDefaultOptions m_allocator;
+  Ort::MemoryInfo m_memoryInfo;
+
+  /// Input/Output names and input shape
+  std::vector<char*> m_inputNames;
+  std::vector<char*> m_outputNames;
+  std::vector<std::vector<int64_t>> m_inputShapes;
+};
+
+class VAEModelSimulation : public NeuralFastSimulation
+{
+ public:
+  VAEModelSimulation(std::array<float, 9> conditionalMeans,
+                     std::array<float, 9> conditionalScales,
+                     float noiseStdDev);
+  ~VAEModelSimulation() = default;
+
+  void run() override;
+  std::array<int, 5> get_channels() override;
+  void set_data(std::array<float, 9>& particle);
+
+ private:
+  std::array<float, 9> scale_conditional_input(const std::array<float, 9>& rawConditionalInput);
+
+  std::array<float, 9> m_conditionalMeans;
+  std::array<float, 9> m_conditionalScales;
+  float m_noiseStdDev;
+  std::vector<float> m_noiseInput;
+  std::array<float, 9> m_particle{};
+  std::vector<Ort::Value> m_modelOutput;
+};
+
+class SAEModelSimulation : public NeuralFastSimulation
+{
+ public:
+  SAEModelSimulation(std::array<float, 9> conditionalMeans, std::array<float, 9> conditionalScales);
```

Comment:
try to use `const &` consistently in all interfaces taking complex objects

Reply:
Fixed in newest commits.